Create simulate-able base class for plant (HVAC?) components

Myoldmopar opened the pull request:
## Overview

A new base class is created for objects that are called to simulate by plant.  The plant data structure is expanded to include a shared pointer to an instance of this base class.  This is shared because the component can be called (Sim_()) at different locations in plant, so the shared pointer simply calls the same Sim_ as before.  Components (right now only pipes) are modified to derive from the base class and work with the virtual methods defined in the base class.  Investigation will be performed to see how far this can be pushed (all HVAC components? other?)

Several things modified here, described in the order they are shown below at the time of this writing.
### DataPlant.hh
- Added a shared pointer to the plant component structure that will point to a "PlantComponent" instance 
- Cleaned up a bunch of unused member constructors in DataPlant
### Pipes.cc
- Modified the Pipes code significantly:
  - Component array is now an array of shared pointers to the LocalPipeData _class_, which derives from the PlantComponent base class
  - _Construction_ is handled by a factory which is a static class method in LocalPipeData, and only constructs a _single_ pipe with a name passed in as the only factory argument
  - The factory has 2 purposes:
    - Instantiate a shared pointer to a LocalPipeData instance and add it to the local component array
    - Return a shared pointer to the same data but polymorphed into the PlantComponent base class, which is stored by the plant for calling later
- Extracted out all the different levels of initialization into separate functions:
  - Every-time initialization
  - One-time initialization
- Extracted out the simulation calculations into a separate simulate() function
### Pipes.hh
- Changed pipe struct into a class that derives from the new PlantComponent base class
- Changed main component array to store shared pointers instead of actual object instances
- Moved functions from global into actual class methods _completely_, although the factory method is still static...on purpose
### PlantComponent.hh
- This is a brand new file, and will be in flux for a little while
- It basically defines the interface (pure virtual base class) for components being called to simulate _from plant_
- It could be generalized to other HVAC components as well, but maybe it is better to keep it a little lower scope
### PlantLoopEquip.hh
- Only modified the pipes section...this is **key** in how I am doing this because it can be done one-by-one until all the components are complete
- Initially this will make it somewhat bulkier, but once everything is done this will be **so** much cleaner
- For pipes, I operate directly on the component pointer in the plant data structure, instead of calling global functions and passing in arguments.  
- Right now, the checks for one-time init and envrn-inits happen as each component is called, but once this whole thing gets pushed to all the components, we won't have to do these checks on each component, and we will finally get a performance boost out of this.
- Even with these checks remaining _as they are now_, we are still getting a miniscule boost because the components are not checking _GetInput?_ every single time Sim\* is called.
### PlantManager.cc
- When the plant is building up the loop topology and assigning type of numbers, etc., it now calls the new pipe factory method also.  This returns the shared pointer to the plant component so that it can be called later (in PlantLoopEquip above).  
## So what's the plan?
- This method is nice because it can be implemented gradually without a one-all breaking of the code/solver.  It also should produce no -- or _extremely_ small -- diffs.
- To add more plant components, the following steps are to be taken:
  - Convert the component class to derive from the PlantComponent base class
  - Extract out the initialization/simulation/construction pieces into the appropriate virtual methods
  - Modify PlantManager so that the compPtr will be created and available
  - Modify PlantLoopEquip to take advantage of this base class
## ToDo:
- [x] Pull out the logic in PlantLoopEquip for the pipes into a small worker function that can be reused
- [ ] Re-implement name verification in the component construction
- This should work out of the box for singly-connected plant components.  Test out for a few types of mroe complex plant components:
  - [ ] Plant to plant (water cooled chiller) -- should work, except that we might need to add an argument to Sim\* to state where we are calling from (which branch/comp/etc.)
  - [ ] Plant to plant to plant (add heat recovery)
  - [ ] Plant to air (chilled water coil) -- we only manage the interface to the plant, but there could be complications...

Myoldmopar committed:
Yet another whack at objectifying the plant;
I think this work finally, successfully, kicks off this work;
Basically it extracts out construction and different levels of initialization to plant management,
 which makes it quicker because these boolean checks arent littered through the code;
Right now it wont be much faster but this will make it possible to move toward improvements;
This also extracts out a factory class to do construction

Myoldmopar committed:
Extract out logic for simulation of new component;
Fixes the issue in at least a good number of the problem files in the first commit.

Myoldmopar commented:
So with the second commit, it looks like every file with a pipe has audit and bnd diff's which are due to the order of how pipes are retrieved. I didn't see any eio diff's. I'll run the rest suite locally to verify, but this could be all clean.

Myoldmopar commented:
Umm, @DeadParrot (or @lefticus maybe), could you assist with the [warning message coming from gcc](https://nrel.github.io/EnergyPlusBuildResults/EnergyPlus-beef29a2738510228cde91cc822c1f2b322911eb-x86_64-Linux-Ubuntu-14.04-gcc-4.8.html)?

Here's the full text on my local machine, release build of this branch:

```
In file included from /.../third_party/ObjexxFCL/src/ObjexxFCL/DimensionSlice.hh:17:0,
                 from /.../third_party/ObjexxFCL/src/ObjexxFCL/ArrayS.hh:19,
                 from /.../third_party/ObjexxFCL/src/ObjexxFCL/Array.hh:19,
                 from /.../third_party/ObjexxFCL/src/ObjexxFCL/Array1.hh:18,
                 from /.../third_party/ObjexxFCL/src/ObjexxFCL/Array1D.hh:18,
                 from /.../third_party/ObjexxFCL/src/ObjexxFCL/Array.functions.hh:17,
                 from /.../src/EnergyPlus/EconomicTariff.cc:6:
/.../third_party/ObjexxFCL/src/ObjexxFCL/IndexRange.hh: In function ‘void EnergyPlus::EconomicTariff::popStack(ObjexxFCL::Array1A<double>, int&)’:
/.../third_party/ObjexxFCL/src/ObjexxFCL/IndexRange.hh:569:3: warning: assuming signed overflow does not occur when assuming that (X + c) < X is always false [-Wstrict-overflow]
   if ( l_ > u ) {
   ^
```

Right now the only thing I can think of is that I am actually _using_ the new push_back method on Array1D, so maybe it is coming out of that usage.

Myoldmopar commented:
Oh, wait, dang, it is happening in the develop branch for me.  And for CI.  Develop isn't all green!! :-1:

Yeah, looks like the new Objexx stuff is when this was introduced.  Any thoughts, @DeadParrot?  ( And nevermind, @lefticus )

DeadParrot commented:
@Myoldmopar This is a [known false positive](http://stackoverflow.com/questions/12984861/dont-understand-assuming-signed-overflow-warning) with `-Wstrict-overflow` that is eliminated in GCC 4.9 and later (which is why I didn't see it in my testing). There is no problem with the code (beyond the universal exposure to overflows any time you change a number).

The tweak in PR #4946 eliminates the warning so adding `-Wno-strict-overflow` won't be necessary for GCC 4.8 for this warning but it could appear again elsewhere so moving on from 4.8 soon-ish might be a good idea (this isn't our first work-around for 4.8 flaws).

Myoldmopar commented:
Thanks @DeadParrot, it's not an urgent matter, just kinda caught me off guard!  I'll merge your fix ASAP and pull it back into this branch.

Myoldmopar commented:
Local test suite: lots of text diffs exactly like CI, and **zero** math/table diffs, big _or_ small.  

@mbadams5 @kbenne @lefticus @DeadParrot @lgentile @mjwitte Maybe we should have a little chat about this next week to see if it is something I/we should pursue further toward a restructure...

Myoldmopar committed:
Merge remote-tracking branch 'origin/develop' into PullLogicFromPlantComponents

Myoldmopar committed:
Updates to component base class, core restructuring of pipe HT code, *not building*

mjwitte commented:
It is premature to discuss this on the iteration call and walk through the concept for everyone?

Myoldmopar committed:
Pipe heat transfer conversion builds...not running yet

kbenne commented on src/EnergyPlus/Pipes.cc:
```diff
@@ -31,363 +35,93 @@ namespace Pipes {
+				thisPipe->compType = TypeOf_Pipe;
+				thisPipe->InletNodeNum = GetOnlySingleNode( cAlphaArgs( 2 ), ErrorsFound, cCurrentModuleObject, cAlphaArgs( 1 ), NodeType_Water, NodeConnectionType_Inlet, 1, ObjectIsNotParent );
+				thisPipe->OutletNodeNum = GetOnlySingleNode( cAlphaArgs( 3 ), ErrorsFound, cCurrentModuleObject, cAlphaArgs( 1 ), NodeType_Water, NodeConnectionType_Outlet, 1, ObjectIsNotParent );
+				TestCompSet( cCurrentModuleObject, cAlphaArgs( 1 ), cAlphaArgs( 2 ), cAlphaArgs( 3 ), "Pipe Nodes" );
```
break?

kbenne commented on src/EnergyPlus/Pipes.cc:
```diff
@@ -31,363 +35,93 @@ namespace Pipes {
+				thisPipe->compType = TypeOf_PipeSteam;
+				thisPipe->InletNodeNum = GetOnlySingleNode( cAlphaArgs( 2 ), ErrorsFound, cCurrentModuleObject2, cAlphaArgs( 1 ), NodeType_Steam, NodeConnectionType_Inlet, 1, ObjectIsNotParent );
+				thisPipe->OutletNodeNum = GetOnlySingleNode( cAlphaArgs( 3 ), ErrorsFound, cCurrentModuleObject2, cAlphaArgs( 1 ), NodeType_Steam, NodeConnectionType_Outlet, 1, ObjectIsNotParent );
+				TestCompSet( cCurrentModuleObject2, cAlphaArgs( 1 ), cAlphaArgs( 2 ), cAlphaArgs( 3 ), "Pipe Nodes" );
```
break?

Myoldmopar replied:
Excellent find @kbenne, but might hold on briefly as I've fixed these and several other things while working on the pipe heat transfer conversion over the weekend and this morning.  

Feel free to review if you want, it just might change _very_ shortly.

kbenne commented on src/EnergyPlus/Pipes.hh:
```diff
@@ -1,129 +1,53 @@
-		int & PipeNum, // Index into pipe structure for name
-		Real64 const MaxVolFlowRate // unused at present time
-	);
+	extern Array1D< std::shared_ptr< LocalPipeData > > LocalPipe; // dimension to number of pipes
```
I know you're still working on this, but why not std::vector or other std container?

jasondegraw replied:
+1 on the use of std containers

Myoldmopar replied:
Oh no! Not this conversation!

mbadams5 replied:
:+1: on the use of std containers as well...

DeadParrot replied:
+1 on std containers as the first choice to consider for, well, containers. Even if the performance is the same with `Array1D` the API of `std::vector` will be more familiar. The biggest issues with `std::vector` for us are the 0-based indexing and familiarity with `reserve` and `shrink_to_fit` to get good performance during growth without wasted RAM.

The 0-based indexing sounds trivial but I predict it will cause problems and bugs, especially as we mix `std::vector` with arrays. A solution that I've used before is to provide a variant of `std::vector` that has 1-based indexing. (It can also fix the lack of bounds checking in debug builds -- some compilers do this now but it isn't in the C++ spec.)

As a side note, `Array1D` just got a vector-like push_back, capacity, reserve, and shrink-to-fit API, as well as iterators, so it can be used as a std-like container with similar performance to `std::vector`. This, plus 1-based indexing, may explain why Edwin stayed with Array1D.

DeadParrot replied:
Wrt containers of smart pointers I will note that I'm finishing up a "collection" framework that is meant to let us work with these sorts of entities more easily. Some of the features of this are:
- Automatic dereferencing of iterators and subscripts to avoid pointer syntax
- Sub-collections are easily generated and lightweight (pointers to elements internally)
- Uniform, simple API
- 1-based indexing (collection(1) is the first element)
- Grow-friendly internal containers (like std::vector) to reduce heap allocations

Sub-collections will be used to improve performance and scalability of performance-critical loops that now loop over all elements but only operate on a subset. This is a big performance hit that gets worse for larger models (big-_O_ complexity for you CS types).

Boost pointer containers and indirect iterators provide some of this but not everything I'm going for and bring in a big dependency and (possibly much) slower builds. I'm happy to provide more details.

jasondegraw replied:
@DeadParrot I hate to be a wet blanket, but 1-based indexing is an absolute non-starter as far as I am concerned. New code should only be using 1-based indexing in extraordinary circumstances and we should be working to get away from it, not adding more. Sorry.

Myoldmopar replied:
@jasondegraw went to the store and bought 0 apple. He then ate said apple.  Now he has empty apples.

Myoldmopar replied:
Also, @jasondegraw lives at address: 0 w street.

Geez, now I'm being a wet blanket @nllong.

OK, in more serious matters, because of the way I am structuring this, I couldn't care less what the indexing is because I am operating on smart pointers to instances of those classes, so the indexing will be m00t.

jasondegraw replied:
@Myoldmopar I did wear 00 as a jersey number back in the day...

Yes, it really shouldn't matter (use ranged for, etc), but I remain steadfastly against 1-based indexing.

DeadParrot replied:
@jasondegraw This is a good debate to have now (or a year ago ;-). I think it is hard to work on a code that is partly 0-based and partly 1-based and, while you may be comfortable with 0-based, I have worked with groups transitioning from Fortran and 0-based is a big source of grief (and bugs). But, happily, the collection framework, like the Arrays, does both:

```
collection(1); // First element
collection[0]; // First element
```

Yes, this can be another source of confusion without good communication, but it does let you go with 1-based for now and switch down the road. And it would take about 3 minutes to make the collections do everything 0-based, so let's get this debated and decided!

jasondegraw replied:
@DeadParrot Sure, but you should know that I'm not convinced that there is a need for a "collection" object either. I'd rather see

```
std::vector<std::shared_ptr<SomeClass>> objects;
```

than something that hides the details. But setting that aside, what is the use case in new code for 1-based indexing? The way I see it, references to objects in the vector (or whatever) should be via a shared pointer and looping should use the ranged-based for or iterators. Is there some other use that I am missing?

DeadParrot replied:
@jasondegraw Wrt indexing, agree on prefering iteration and range-based for loops. But we have a lot of index-based code and (non-container) multidimensional arrays need it. When you are doing array computations indexing is more natural and efficient (you don't want to be iterating over multiple iterators in parallel). For containers, the use of indexing is less compelling but there are times when you need to move through multiple containers in parallel so indexing is more efficient there, too. I.e., we still need indexing. If we want to aim to be all modern and stuff how about a guideline that all-new code use 0-based container indexing and updates to existing code can use 1-based (when available). Having both is one advantage of wrapping `std::vector` in a collection.

There are a lot of other reasons for a collection system (or else [Boost Pointer Containers](http://www.boost.org/doc/libs/1_58_0/libs/ptr_container/doc/ptr_container.html) wouldn't exist). Not for local algorithm containers necessarily, but for the major model components they are a win for our developer pool (who are not as comfortable with C++ as you are):
- Containers of values and containers of pointers work completely differently: having a common API for them helps our developers.
- Is the pointer syntax of polymorphic containers a leaked artifact of their implementation? (Yes: anti-OO: ergo [Boost indirect_iterator](http://www.boost.org/doc/libs/1_58_0/libs/iterator/doc/indirect_iterator.html))
- What happens when you run std::sort on the vector of shared_ptr?!?
- Does `std::vector` guarantee to bounds check lookups in debug builds? (No.)
- Can developers figure out how to get the windows subset of surfaces and even for those that can isn't that something we should package up with templatety goodness? (Yes, this could be a function if we don't have collection templates but it is nice to have a sub-collection hold its membership predicate for validation, refresh, ...)

I do want to get feedback on the collections framework from the team but it should consider the average, as well as expert, developer. I can share the design doc and code/examples with anyone interested.

[Another discussion, but `uniqe_ptr`, not `shared_ptr`, should be the default way to hold polymorphic types: only pay for shared ownership if you really need it.]

jasondegraw replied:
Well played, @Myoldmopar, well played. Outdate the diff so I can't find the conversation...

@DeadParrot Sorry, but I don't agree with much of that. I agree that for computations, indexing is often the way to go, but this is not that case. I would be fine with a rule for new code, but I wouldn't phrase it in terms of indexing. The rule should be standard containers for new code. Standard containers are, well, standard, and there's a wealth of information available on how to use them. Using special wrappers might help current developers, but presents a barrier to new developers. In my opinion that is a major net loss. Here are my responses to your list:
- I'm not sure if I understand your point, but why should containers of values work the same as containers of pointers? Seems to me that they're different.
- Yes, but that's OK, C++ is a bit of a mess
- Comparison function?!?
- Bounds checking only matters if there's a lot of indexing going on and I think I've been clear what I think about that.
- If they're using standard containers, then yes, that's something that Google can solve. I'd prefer a function over an entire object that'll require developers to sift through code to determine what it does.

Apologies for rejecting everything you're saying, but I don't see this as an average vs expert thing. In any case, I'd be happy to look at a design doc and code.

Myoldmopar committed:
Minor tweaks, and the pipe ht file runs...not yet verified outputs

kbenne commented on src/EnergyPlus/PlantComponent.hh:
```diff
@@ -0,0 +1,20 @@
+{
+public:
+	std::string name;
+	int compType;
```
This is curious to me.  In some places I see types identified by string.  (Like up in the factory for pipe.)  In other places there is string and int.  (Like here https://github.com/NREL/EnergyPlus/pull/4945/files#diff-ae587fad7fd53d7ab16d3d93e273bcefR182 and here https://github.com/NREL/EnergyPlus/pull/4945/files#diff-d8cdb0f9f5f04513135a9405b38ec78dR11.  I think I understand the drivers.  When you're coming from idf input you are talking in terms of strings, but when you get inside you probably want the int for fast comparison.  Polymorphism might reduce the need for determining type, but I imagine it will still be around a fair bit.  How about unifying these into a new type like [IddObjectType](https://github.com/NREL/OpenStudio/blob/develop/openstudiocore/src/utilities/idd/IddEnums.hpp#L105) that has string integer bound together?

Myoldmopar replied:
That's a great idea @kbenne looking forward.  Whenever we do talk about this we'll keep it in mind.  One thing I want to balance is how this current approach is set up to work perfectly within the existing structures while still making big strides on a component-by-component level.  If the IddObjectType can be used in this same fashion; awesome.  If not, I'm not sure we'll pull it in until we've pushed this along quite a ways.

Myoldmopar commented:
Just pushed the changes to the PipeHeatTransfer code.  It works perfectly...no diffs except the same bnd and audit order differences.  This is going swimmingly.

The pipe heat transfer code is much larger and so it didn't come out as super clean as the regular pipes, so I'll make a clean up pass on it now that it is running correctly.

Myoldmopar committed:
Clean up gobs of using statements and other stuff

Myoldmopar commented:
@mjwitte I am OK discussing this on the call.  The extent of that discussion will depend on how far I get today.  I am going to work on a more complex object, probably a chiller, to make sure there aren't any surprises moving forward.

mjwitte commented:
@Myoldmopar Looks like Modelica is already on the agenda so we probably won't have time for that.  Perhaps the next sizing call would be a good time to do an initial walk-through.

Myoldmopar commented:
Yeah, I'll have it nailed down a little better by then...even so far since
I commented I've got the chillers underway and it looks like it is going to
be pretty nice.

Myoldmopar committed:
Intermediate working copy of chiller re-do
Moved chiller files to new subdir;
Moved plant location to its own file to resolve circular dependency;
Initial Modifications to plant component interface to handle multiply connected components

Myoldmopar committed:
Get the new chiller code files building...methods arent implemented yet

Myoldmopar committed:
Const COP chiller initialization code now all in place and building.  No testing yet.

Myoldmopar committed:
Code pasted in but commented still...builds but not tested yet;
The calc function will be the most fun, lots of code to pull out probably;
then handling the previous module level variables....

Myoldmopar committed:
Conversion completed and the code runs.  Running diff tests next.

Myoldmopar committed:
Changes to API for initLoopEquip capabilities;
Chillers now size properly with the calls to SimPlantEquip from InitializeLoops;
The API may be nearing stability after a look-over from some others

EnergyArchmage commented:
I think this looping over GetObjectItem I see in the factory methods here are clumsy. 
For example,
        for ( int ChillerNum = 1; ChillerNum <= NumConstCOPChillers; ++ChillerNum ) {
            InputProcessor::GetObjectItem( cCurrentModuleObject, ChillerNum, cAlphaArgs, NumAlphas, rNumericArgs, NumNums, IOStat, _, lAlphaFieldBlanks, cAlphaFieldNames, cNumericFieldNames );
            if ( objectName == DataIPShortCuts::cAlphaArgs( 1 ) ) {
                found = true;
            } else {
                continue;
            }

 It would be better to get the index from IP and then call GetObjectItem with that, using the function InputProcessor::GetObjectItemNum(cCurrentModuleObject, objectName ); .  This allows getting rid of the for loop you have as it will generate a lot of extra calls to GetObjectItem that aren't useful.

Myoldmopar committed:
Intermediate commit; still debugging

Myoldmopar committed:
Massive merge from develop; including tree conflicts; fingers crossed

Myoldmopar committed:
Fix build errors; potentially up-to-date and working again

Myoldmopar commented:
During the tree conflict, the pipe files didn't understand there was a rename.  As I move this toward completion, I may end up re-starting with a fresh branch and doing all the changes in one commit at once.  This will allow me to easily make sure I get all the renames and mods tracked properly, and get rid of any noise from intermediate commits.

Myoldmopar committed:
Merge develop, resolve minor conflict; remove build warning

Myoldmopar commented:
OK, develop merged in again.  Next step is running develop and this in simultaneous gdb sessions to figure out where the diffs are coming from.  If I can't figure it out in reasonable time, I might pull back the Chiller restructuring and put it back the way it was.

Myoldmopar committed:
Eliminate silly build warnings

nrel-bot-3 commented:
@Myoldmopar @lgentile it has been 14 days since this pull request was last updated.

nrel-bot-3 commented:
@Myoldmopar @lgentile it has been 14 days since this pull request was last updated.

nrel-bot-3 commented:
@Myoldmopar @lgentile it has been 14 days since this pull request was last updated.

Myoldmopar commented:
This has been superceded by #5426 and will be closed without merge.  But I'll leave it hanging around because I am using the diff view of this PR as I work on the other PR.

Myoldmopar commented:
I'm closing this, I'll leave the branch around temporarily.
